kaspresso/src/main/kotlin/com/kaspersky/kaspresso/docloc/metadata/extractor/UiMetadataExtractor.kt

```diff
@@ -0,0 +1,96 @@
+
+import android.app.Activity
+import androidx.test.uiautomator.By
+import androidx.test.uiautomator.StaleObjectException
+import androidx.test.uiautomator.UiDevice
+import com.kaspersky.kaspresso.device.activities.Activities
+import com.kaspersky.kaspresso.docloc.metadata.LocalizedString
+import com.kaspersky.kaspresso.docloc.metadata.Metadata
+import com.kaspersky.kaspresso.docloc.metadata.Window
+import com.kaspersky.kaspresso.logger.UiTestLogger
+
+internal class UiMetadataExtractor(
+    private val uiDevice: UiDevice,
+    private val activities: Activities,
+    private val logger: UiTestLogger,
+) : MetadataExtractor {
+
+    private val metadataExtractorHelper = MetadataExtractorHelper()
+
+    override fun getMetadata(): Metadata {
+        val activity = activities.getResumed() ?: throw RuntimeException("Failed to get current activity")
+        return createMetadata(activity)
+    }
+
+    private fun createMetadata(activity: Activity): Metadata {
+        val localizedStrings = metadataExtractorHelper.resolveAmbiguous(
+            getLocalizedStrings()
+        )
+        val window = activity.window.decorView.run {
+            Window(
+                left,
+                top,
+                width,
+                height,
+                localizedStrings
+            )
+        }
+        return Metadata(window)
+    }
+
+    private fun getLocalizedStrings(): List<LocalizedString> {
+        uiDevice.setCompressedLayoutHierarchy(false)
+        val objectsWithText = uiDevice.findObjects(By.enabled(true))
+            .filterNotNull()
+            .filter {
+                try {
+                    (it.resourceName != null && it.children.any { !it.text.isNullOrBlank() }) ||
+                            (!it.text.isNullOrBlank() && it.resourceName != null)
+                } catch (ex: StaleObjectException) {
+                    logger.e("UiMetadataExtractor::getLocalizedStrings() - Error while loading object")
+                    false
+                }
+            }
+
+        val localizedStrings = mutableListOf<LocalizedString>()
+        for (obj in objectsWithText) {
+            try {
+                val resName = obj.resourceName ?: continue
+                if (!obj.text.isNullOrBlank()) {
+                    val coords = obj.visibleBounds
+                    localizedStrings.add(
+                        LocalizedString(
+                            text = obj.text,
+                            locValueDescription = resName,
+                            left = coords.left,
+                            top = coords.top,
+                            width = coords.width(),
+                            height = coords.height(),
+                        )
+                    )
+                } else {
+                    obj.children
+                        .filter { !it.text.isNullOrBlank() }
+                        .forEach {
+                            val coords = obj.visibleBounds
```

Comment:
same code can be moved to a separate method

Reply:
You found a possible bug. When we iterate through the children we better use the child coordinates